### PR TITLE
Windows: Add UWP target stub [Flutter#14697]

### PR DESCRIPTION
**Pull request by clarkezone**
## Description

This is the first in a series of patches to add support for a Windows UWP target, embedder implementation and runner per https://github.com/flutter/flutter/issues/14967.

This change adds a new build target which is selected with target os `winuwp` for example `python .\flutter\tools\gn --runtime-mode=debug --unoptimized --winuwp` and will build a Windows binary flutter_windows_winuwp.dll.  The implementation is stub only but the important aspect of this change is that all win32 specifics are now compiled behind an `#ifdef` which is only active for the win32 target.  Once this PR has landed, the intent is to enable the target in CI to give some level of protection for the UWP target as new win32 functionality and os calls are added.

This change cannot be landed until a small 5 line patch is made to dart to support target_os == "winuwp" which will be opened in due course and referenced here.  Until that has happened this change will not build successfully in CI and should be consider WIP although it is ready for review now.

## Related Issues

https://github.com/flutter/flutter/issues/14967

## Tests

Since this change is a stub, testing has not been added in anything but placeholder form.

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [contributor guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [C++, Objective-C, Java style guides] for the engine.
- [x] I read the [tree hygiene] wiki page, which explains my responsibilities.
- [ ] I updated/added relevant documentation.
- [x] All existing and new tests are passing.
- [x] I am willing to follow-up on review comments in a timely manner.

## Reviewer Checklist

- [ ] I have submitted any presubmit flakes in this PR using the [engine presubmit flakes form] before re-triggering the failure.

## Breaking Change

Did any tests fail when you ran them? Please read [handling breaking changes].

- [ ] No, no existing tests failed, so this is *not* a breaking change.
- [ ] Yes, this is a breaking change. *If not, delete the remainder of this section.*
   - [ ] I wrote a design doc: https://flutter.dev/go/template *Replace this with a link to your design doc's short link*
   - [ ] I got input from the developer relations team, specifically from: *Replace with the names of who gave advice*
   - [ ] I wrote a migration guide: https://flutter.dev/go/breaking-changes-template *Replace this with a link to a pull request that adds the migration guide to https://flutter.dev/docs/release/breaking-changes*

[issue database]: https://github.com/flutter/flutter/issues
[contributor guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[testing the engine]: https://github.com/flutter/flutter/wiki/Testing-the-engine
[C++, Objective-C, Java style guides]: https://github.com/flutter/engine/blob/master/CONTRIBUTING.md#style
[CLA]: https://cla.developers.google.com/
[tree hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[engine presubmit flakes form]: https://forms.gle/Wc1VyFRYJjQTH6w5A

### CODE REVIEW

**Review by stuartmorgan — CHANGES_REQUESTED** [review]
🤦 I just discovered that this has been sitting here as a draft review for over a week. I thought it was waiting for you to update based on feedback.

File: BUILD.gn
Comment (clarkezone) [reply BUILD.gn:153]: Addressed in next commit

File: shell/platform/windows/BUILD.gn
Comment (clarkezone) [reply shell/platform/windows/BUILD.gn:24]: Fixed in next commit

File: shell/platform/windows/BUILD.gn
Comment (clarkezone) [reply shell/platform/windows/BUILD.gn:26]: I've updated the name and value but moving it to build/config/win/BUILD.gn means that it's a dependent change in buildroot that would have to land ahead of this change.  I would rather land without moving since there will have to be a subsequent change to buildroot with other changes and i'd rather not complicate the dependencies further.  I've added a TODO to that effect, Please reopen if you disagree.

File: shell/platform/windows/BUILD.gn
Comment (clarkezone) [reply shell/platform/windows/BUILD.gn:166]: went with the approach of conditionalizing single sources per your next comment.  Fixed in next commit.

File: shell/platform/windows/BUILD.gn
Comment (clarkezone) [reply shell/platform/windows/BUILD.gn:140]: Fixed in next commit

File: shell/platform/windows/BUILD.gn
Comment (clarkezone) [reply shell/platform/windows/BUILD.gn:214]: Fixed in next commit.

File: shell/platform/windows/flutter_windows.cc
Comment (clarkezone) [reply shell/platform/windows/flutter_windows.cc:25]: Fixed in next commit.

File: shell/platform/windows/flutter_windows.cc
Comment (clarkezone) [reply shell/platform/windows/flutter_windows.cc:147]: Fixed in next commit.

File: shell/platform/windows/BUILD.gn
Comment (clarkezone) [reply shell/platform/windows/BUILD.gn:40]: I don't understand this.

File: shell/platform/windows/flutter_windows.cc
Comment (clarkezone) [reply shell/platform/windows/flutter_windows.cc:79]: I don't understand this feedback, appols.

File: shell/platform/windows/flutter_windows_engine.cc
Comment (clarkezone) [reply shell/platform/windows/flutter_windows_engine.cc]: Appols, please can you be more specific?

File: shell/platform/windows/flutter_windows.cc
Comment (clarkezone) [reply shell/platform/windows/flutter_windows.cc:84]: Ifdefs in this file including this example are around flat functions so i'm not sure how exactly to introduce a subclass.  Are you suggesting including a target specific version of the file flutter_windows.cc implementation (flutter_windows_32.cc / flutter_windows_winuwp.cc) which isn't a subclass but would remove the need for the ifdefs?

File: shell/platform/windows/BUILD.gn
Comment (stuartmorgan) [reply shell/platform/windows/BUILD.gn]: See my comment on your new buildroot PR; there shouldn't be a need to manually add a new "this is UWP" config to individual targets if you change it as described there.

File: shell/platform/windows/flutter_windows.cc
Comment (stuartmorgan) [reply shell/platform/windows/flutter_windows.cc:85]: 🤦 Somehow I thought this was flutter_engine_windows when leaving that comment. But I think having `flutter_windows.cc` with common code, and `flutter_windows_win32.cc` and `flutter_windows_winuwp.cc` with the platform-specific parts, would be better than ifdefing a bunch of entire function bodies.

File: shell/platform/windows/flutter_windows.cc
Comment (stuartmorgan) [reply shell/platform/windows/flutter_windows.cc:80]: I just realized this part is pure placeholder; I was saying that part of the code above was relevant to UWP even though not all of it was. But since there's no implementation at all yet, never mind.

File: shell/platform/windows/BUILD.gn
Comment (clarkezone) [reply shell/platform/windows/BUILD.gn]: Addressed.

File: shell/platform/windows/flutter_windows.cc
Comment (clarkezone) [reply shell/platform/windows/flutter_windows.cc]: Done.

File: shell/platform/windows/flutter_windows_engine.cc
Comment (clarkezone) [reply shell/platform/windows/flutter_windows_engine.cc]: Anything to do here?

**Comment by clarkezone** [comment]
One remaining clarifying question.  This change is now dependent on a buildroot update: https://github.com/flutter/buildroot/pull/409

File: shell/platform/windows/flutter_windows_engine.cc
Comment (stuartmorgan) [reply shell/platform/windows/flutter_windows_engine.cc]: Sorry I missed this. In particular, I think we should make an abstract TaskRunner class that the current Win32 version inherits from, and with a temporarily-no-op UWP subclass, and then most of the ifdefs about task runners in this file go away (in fact, if we declare a factory method in the abstract class's header, and implement it in each of the subclass files, then we don't need any ifdefs for task runner at all I think.)

Comment (clarkezone) [reply shell/platform/windows/flutter_windows_engine.cc]: I have added an abstract taskrunner, made the win32 version inherit and add the winrt version as well (bonus.. it actually works ;-)).  However, I'm not sure how to implement the factory pattern in a way that would eliminate the ifdef from flutter_windows_engine.  Since c++ doesn't support virtual static inheritance i assume you are implying using a template function for the factory?  If yes, that will not eliminate the need for an ifdef at the calling site.  Also, note that there is another ifdef in that file so even if the factory is possible, that file won't be clean.  Moving the ifdefs in question up into the caller will merely add ifdefs to a file that doesn't have any currently.  So not quite sure what to do about the last aspect of the request here.

Comment (clarkezone) [reply shell/platform/windows/flutter_windows_engine.cc]: Please reopen if you have more thoughts.

Comment (stuartmorgan) [reply shell/platform/windows/flutter_windows_engine.cc]: I was envisioning a more low-tech solution:
- Declare a freestanding `CreateTaskRunner` function in task_runner.h
- Implement that function twice, once in the Win32 .cc file, and once in the WinRT .cc file

Since only one of the files will be linked, only one implementation will be present.

File: shell/platform/windows/win32_task_runner.h
Comment (stuartmorgan) [inline shell/platform/windows/win32_task_runner.h:26]: We should really be calling these classes things like `TaskRunnerWin32`/`TaskRunnerWinRt` rather than the reverse. Let's fix that for this class in this PR (since it's adding a new file with the naming we should move away from), and I can do a follow-up to clean up the others after it lands.

File: BUILD.gn
Comment (stuartmorgan) [inline BUILD.gn:152]: We can reference https://github.com/flutter/flutter/issues/70197 here.

File: shell/platform/windows/BUILD.gn
Comment (stuartmorgan) [inline shell/platform/windows/BUILD.gn:78]: Can we be consistent about winrt vs winuwp, or is the distinction meaningful? I.e., is there a scenario where this file would be used for non-UWP? (If I look up the CoreWindow code there I get docs in the UWP section of docs.microsoft.com.)

File: shell/platform/windows/flutter_windows_view.h
Comment (stuartmorgan) [inline shell/platform/windows/flutter_windows_view.h:25]: Conditional includes need to be pulled out to their own block.

File: shell/platform/windows/flutter_windows_view.cc
Comment (stuartmorgan) [inline shell/platform/windows/flutter_windows_view.cc:43]: We should be able to make this one have a generic interface too.

File: shell/platform/windows/winrt_task_runner.cc
Comment (stuartmorgan) [inline shell/platform/windows/winrt_task_runner.cc:36]: The current implementation isn't using the target time, so I'm confused why this is "consider"; presumably the current implementation is wrong?

File: shell/platform/windows/winrt_task_runner.h
Comment (stuartmorgan) [inline shell/platform/windows/winrt_task_runner.h:31]: These should be the function-defining superclass's name rather than the function name.

File: tools/gn
Comment (stuartmorgan) [inline tools/gn:144]: Can you link to the PR?

**Review by stuartmorgan — CHANGES_REQUESTED** [review]
Mostly just some more small structural cleanup so we're not adding more technical debt; overall seems like a good foundation.

File: tools/gn
Comment (clarkezone) [reply tools/gn:144]: I haven't yet opened the PR so maybe we should revise the comment

File: shell/platform/windows/winrt_task_runner.cc
Comment (clarkezone) [reply shell/platform/windows/winrt_task_runner.cc:36]: Yes it's wrong.

File: shell/platform/windows/BUILD.gn
Comment (clarkezone) [reply shell/platform/windows/BUILD.gn:78]: Agree that things aren't 100% consistent, i've tried to use winrt to refer to classes that exclusively call winrt API's and not win32.  There will be cases in the future where WinRT API's are called from the desktop version of some classes hence maybe it would be clearer to use UWP

Comment (stuartmorgan) [reply shell/platform/windows/BUILD.gn:78]: I'll standardize on UWP; I think that since we generally call them the Win32 embedding and the UWP embedding, we should use those labels for files that are only for one or the other (even if we potentially end up with things like "win32" files that use WinRT--but hopefully over time migration to WinRT would come with unforking classes to use a shared Reunion API?)

Comment (clarkezone) [reply shell/platform/windows/BUILD.gn:78]: Sounds good.  Yes, that should be the case

File: tools/gn
Comment (stuartmorgan) [reply tools/gn:144]: I'm just going to remove the comment; this seems like something we should probably set in a buildroot .gni file based on target_os. We'll see whether there's a good hook point for that once the ANGLE change is in.

Comment (clarkezone) [reply tools/gn:144]: SG

**Review by stuartmorgan — APPROVED** [review]
I've addressed the comments from my last round of feedback.

@cbracken Since I've taken over the last part of this PR, could you review? I've made significant enough changes that it should get an approval from someone other than me.

**Comment by clarkezone** [comment]
@cbracken do you think you might get to this relatively soon?  I'm holding off on the next UWP change until this is merged and I can cleanly rebuild the UWP prototype on ToT.

**Review by cbracken — APPROVED** [review]
LGTM with one minor question/suggestion.

**Comment by clarkezone** [comment]
Awesome, thx, I’ll address that, update to ToT and submit tomorrow.  Thx so much Chris!

Sent from Mail<https://go.microsoft.com/fwlink/?LinkId=550986> for Windows 10 on ARM

**Pull request merged**
